builtin/credential/radius/path_login.go

```diff
@@ -0,0 +1,182 @@
+package radius
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"layeh.com/radius"
+
+	"github.com/hashicorp/vault/helper/policyutil"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+func pathLogin(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: `login/(?P<username>.+)`,
+		Fields: map[string]*framework.FieldSchema{
+			"username": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "Username to be used for login.",
+			},
+
+			"password": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "Password for this user.",
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.UpdateOperation: b.pathLogin,
+		},
+
+		HelpSynopsis:    pathLoginSyn,
+		HelpDescription: pathLoginDesc,
+	}
+}
+
+func (b *backend) pathLogin(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	username := d.Get("username").(string)
+	password := d.Get("password").(string)
+
+	if username == "" {
+		return logical.ErrorResponse("username cannot be emtpy"), nil
+	}
+
+	if password == "" {
+		return logical.ErrorResponse("password cannot be emtpy"), nil
+	}
+
+	policies, resp, err := b.RadiusLogin(req, username, password)
+	// Handle an internal error
+	if err != nil {
+		return nil, err
+	}
+	if resp != nil {
+		// Handle a logical error
+		if resp.IsError() {
+			return resp, nil
+		}
+	}
+
+	resp.Auth = &logical.Auth{
+		Policies: policies,
+		Metadata: map[string]string{
+			"username": username,
+			"policies": strings.Join(policies, ","),
+		},
+		InternalData: map[string]interface{}{
+			"password": password,
+		},
+		DisplayName: username,
+		LeaseOptions: logical.LeaseOptions{
+			Renewable: true,
+		},
+	}
+	return resp, nil
+}
+
+func (b *backend) pathLoginRenew(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	var err error
+
+	cfg, err := b.Config(req)
+	if err != nil {
+		return nil, err
+	}
+
+	username := req.Auth.Metadata["username"]
+	password := req.Auth.InternalData["password"].(string)
+
+	var resp *logical.Response
+	var loginPolicies []string
+	var user *UserEntry
+	if cfg.ReauthOnRenew {
```

Comment:
That's a good question. The answer might be yes; that during initial authentication 2FA is used and the user is given a certain time period for the token to be valid, and 2FA should be required for it to be valid again. The default TTL could simply be set to the max TTL to keep the user experience the best.

In the longer term it might be nice to allow renew to accept a limited set of options during the request, specifically for 2FA handling. However, we have longer term plans to greatly overhaul 2FA which will make this easier so handling it locally here isn't the right way to go.

So, yeah, I think it should fail. But it would be good to document this scenario and potentially suggest setting a high default TTL in that case, especially if users are going to be getting dynamic secrets that will get revoked when the token expires.

@vishalnayak @briankassouf What do you think?